### PR TITLE
implement `create_honeytoken` method

### PR DESCRIPTION
**Pull request by iXioN**
Add create honeytoken feature with marshmallow models

### CODE REVIEW

File: changelog.d/20230425_100744_antonin.lacombe_implement_honeytoken_creation.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+
+### Added
+
+- Add `create_honeytoken` in client .
```
Comment (agateau-gg) [inline changelog.d/20230425_100744_antonin.lacombe_implement_honeytoken_creation.md:4]: nitpicking:
```suggestion
- Add `GGClient.create_honeytoken()` method.
```

File: pygitguardian/models.py [hunk 2]
```diff
@@ -485,6 +485,69 @@ def __repr__(self) -> str:
...
+        }
+    """
+
+    SCHEMA = HoneytokenResponseSchema()
+
+    def __init__(self, **kwargs: Any) -> None:
+        super().__init__()
+        for key, value in kwargs.items():
+            if not hasattr(self, key):
+                setattr(self, key, value)
```
Comment (agateau-gg) [inline pygitguardian/models.py:545]: I am not fond of using kwargs this way because users or tools like mypy cannot know which arguments are valid. I know it's tedious to type, but I'd prefer it if the method explicitly listed its arguments.

Alternatively, you can use the approach from `iac_models.py` which uses `marshmallow_dataclass` to reduce much of the boilerplate.

File: tests/test_client.py [hunk 3]
```diff
@@ -518,7 +524,7 @@ def test_extra_headers(
     request_mock: Mock,
     client: GGClient,
     session_headers: Any,
-    extra_headers: Optional[Dict[str, str]],
+    extra_headers: Dict[str, str] | None,
```
Comment (agateau-gg) [inline tests/test_client.py:527]: We can't do this: we must support Python 3.7.

Besides, it's better to do these kind of changes in a separate PR, using a tool like pyupgrade to update all the code at once.

**Review by agateau-gg — CHANGES_REQUESTED** [review]

File: tests/test_client.py [hunk 4]
```diff
@@ -646,3 +652,75 @@ def test_versions_from_headers(request_mock: Mock, client: GGClient, method):
...
+
+@patch("requests.Session.request")
+def test_create_honeytoken_error(
+    request_mock: Mock,
+    client: GGClient,
+):
+    """
+    GIVEN a ggclient
+    WHEN calling create_honeytoken with parameters
+    THEN the parameters are passed in the request and the returned honeytoken use the parameters
```
Comment (amascia-gg) [inline tests/test_client.py:709]: this line should be about the error

**Review by amascia-gg — CHANGES_REQUESTED** [review]

File: pygitguardian/models.py [hunk 5]
```diff
@@ -485,6 +485,99 @@ def __repr__(self) -> str:
...
+        triggered_at,
+        revoked_at,
+        open_events_count,
+        type_,
+        creator_id,
+        revoker_id,
+        creator_api_token_id,
+        revoker_api_token_id,
+        token,
+        tags,
```
Comment (agateau-gg) [inline pygitguardian/models.py:557]: These are missing type-hinting. Did you try the marshmallow_dataclass approach? It can save you some typing :)

**Review by agateau-gg — CHANGES_REQUESTED** [review]

File: pygitguardian/models.py (hunk 5 above)
Comment (iXioN) [reply pygitguardian/models.py:557]: no sure to make optional or and timezone aware datetime. And I don't want to spend ages in doc as this already works. I will just add the typing here.

**Review by agateau-gg — APPROVED** [review]
Looks good now, thanks!

File: pygitguardian/models.py (hunk 5 above)
Comment (agateau-gg) [reply pygitguardian/models.py]: makes sense.
